File: Introductory Problems/cpp/AppleDivision.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long

int n = 0;
vector<ll> wts;

// bruteforce
int findMinDiff(int n, vector<ll> wts, ll totalWt) {
    ll minDiff = LLONG_MAX, basket1Weight = 0;
    
    for (ll i = 0; i < (ll) pow(2, n); i++) {
        
        string ithPossibilty = bitset<20>(i).to_string();
        ithPossibilty = ithPossibilty.substr(20-n);

        basket1Weight = 0;
        
        for (int j = 0; j < n; j++) {
            if (ithPossibilty[j] == '1') {
                basket1Weight += wts[j];
            }
        
        }
        minDiff = min(abs(totalWt - 2*basket1Weight), minDiff);
    }

    return minDiff;
}

// recursion
ll findMinDiffR(int index, ll crate1, ll crate2) {
    // base case
    if (index == n) {
        return abs(crate1 - crate2);
    }

    // faith
    return min(
        findMinDiffR(index + 1, crate1 + wts[index], crate2),
        findMinDiffR(index + 1, crate1, crate2 + wts[index])
    );
}

// main
int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);
 
    int temp;
    ll totalWt = 0;
    
    cin >> n;
    
    for (int i = 0; i < n; i++) {
        cin >> temp;
        wts.push_back(temp);
        totalWt += temp;
    }
    
    // solve: bruteforce
    //cout << findMinDiff(n, wts, totalWt);

    // solve: recursion
    cout << findMinDiffR(0, 0, 0);

    return 0;
}

File: Introductory Problems/java/AppleDivision.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class AppleDivision {

    private static int n;
    private static Long minDiff = Long.MAX_VALUE;
    private static List<Long> wts = new ArrayList<Long>();

    private static Long minDifference(int index, Long crate1, Long crate2) {
        // base case
        if (index == n) {
            return Math.abs(crate1 - crate2);
        }

        // faith
        return Math.min(
            minDifference(index + 1, crate1 + wts.get(index), crate2),
            minDifference(index + 1, crate1, crate2 + wts.get(index))
        );
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();

        for(int i=0; i < n; i++) {
            wts.add(sc.nextLong());
        }
        sc.close();

        // using recursion
        minDiff = minDifference(0, 0L, 0L);

        // print the result
        System.out.println(minDiff);

        return;
    }

}
